### ISSUE TITLE
postcss-color-functional-notation unexpectedly not working

### ISSUE THREAD

**Issue by satouriko** [csstools/postcss-plugins#978]
### Bug description

Hi there, I found that with the default `postcss-preset-env`, case
```css
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87 107 149 / var(--tw-text-opacity));
}
```

will be downgrade by `custom-properties` to
```css
.text-link {
  --tw-text-opacity: 1;
  color: #576b95;
  color: rgb(87 107 149 / var(--tw-text-opacity));
}
```

and will not further downgrade by `postcss-color-functional-notation` to
```css
.text-link {
  --tw-text-opacity: 1;
  color: #576b95;
  color: rgba(87,107,149,var(--tw-text-opacity));
}
```

after postcss-color-functional-notation@5.0.2 because of https://github.com/csstools/postcss-plugins/pull/835

This makes users between the gap (e.g Chorme 49~64) cannot display color correctly, and there seems no option to override

https://caniuse.com/css-variables
https://caniuse.com/?search=space-separated

### Source CSS

```css
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87,107,149,var(--tw-text-opacity));
}
```

### Expected CSS

```css
.text-link {
  --tw-text-opacity: 1;
  color: #576b95;
  color: rgba(87,107,149,var(--tw-text-opacity));
}
```

### Actual CSS

```css
.text-link {
  --tw-text-opacity: 1;
  color: #576b95;
  color: rgb(87 107 149/var(--tw-text-opacity));
}
```

### Does it happen with `npx @csstools/csstools-cli <plugin-name> minimal-example.css`?

N/A

### What plugin are you experiencing this issue on?

PostCSS Color Functional Notation

### Plugin version

5.0.2

### Node Version

v16.15.1

### Validations

- [X] Follow our [Code of Conduct](https://github.com/csstools/postcss-plugins/blob/master/CODE_OF_CONDUCT.md)
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### Would you like to open a PR for this bug?

- [ ] I'm willing to open a PR

**Comment by romainmenke** [comment]
Hi @satouriko,  
Thank you for reaching out about this!

Can you share your PostCSS config?

What I see in the [playground](https://preset-env.cssdb.org/playground/) :

```css
/* source */
.text-link { 
  --tw-text-opacity: 1; 
  color: rgba(87,107,149,var(--tw-text-opacity));
}

/* becomes */
.text-link { 
  --tw-text-opacity: 1; 
  color: rgba(87,107,149,1); 
  color: rgba(87,107,149,var(--tw-text-opacity));
}
```


```js
const postcssPresetEnv = require('postcss-preset-env'); // v8.4.1

module.exports = {
	plugins: [
		postcssPresetEnv({
			"browsers": [
				"> 0.2% and not dead"
			]
		})
	]
}
```


-------

Where does `color: #576b95;` come from?
This is not from `postcss-custom-properties`?

**Comment by satouriko** [comment]
Hi @romainmenke, sorry for the trouble of understanding. `color: #576b95;` comes from the optimizer. Let me express it again, with the default configuration (`postcssPresetEnv()`)

```css
/* source */
.text-link { 
  --tw-text-opacity: 1; 
  color: rgb(87 107 149 / var(--tw-text-opacity))
}

/* becomes */
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87, 107, 149, 1);
  color: rgb(87 107 149 / var(--tw-text-opacity))
}
```

The problem is the same, `rgba(87, 107, 149, 1)` and `#576b95` are just equivalent. `rgba(87,107,149,var(--tw-text-opacity))` won't be generated and users falling to the gap (e.g Chorme 49~64) will have a downgraded experience.

**Comment by romainmenke** [comment]
Hi @satouriko,

No trouble at all :)

Which version of `postcss-preset-env` do you currently have?

**Comment by satouriko** [comment]
8.4.1

**Comment by romainmenke** [comment]
I think I am on to something : 

```css
.test-property-with-var {
	--opacity: 1;
	color: rgba(87 107 149 / var(--opacity));
}
```

becomes :

```css
.test-property-with-var {
	--opacity: 1;
	color: rgba(87, 107, 149, 1);
}

@supports (color: rgba(0 0 0 / 0)) and (top: var(--f)) {
.test-property-with-var {
	color: rgba(87 107 149 / var(--opacity));
}
}
```

This does not make it possible to have `color: rgba(87, 107, 149, var(--opacity))` for Chrome 49~64 but every visitor will have a good fallback value.

**Comment by satouriko** [comment]
Wouldn't that end up with a wrong render result if --opacity is overridden by another selector? Since we have color: rgba(87, 107, 149, 1) instead of color: rgba(87, 107, 149, var(--opacity)). Custom properties' document says it only takes css variables definitions in html and :root and the same block.

Currently I find the only way to work around this is to set custom properties to { preserve: false } which would turn off the polyfill if not confident enough.

**Comment by romainmenke** [comment]
Couple of thoughts :)

If you want `rgba(87 107 149 / 1)` then you should write that and not `rgba(87 107 149 / var(--opacity))`. If Tailwind makes that impossible you should open an issue with them :)

---------------

> Custom properties' document says it only takes css variables definitions in html and :root and the same block.

The plugin only looks within the same block when `{ preserve: true }`.

```css
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87, 107, 149, var(--tw-text-opacity));
}
```

becomes :

```css
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87, 107, 149, 1);
  color: rgba(87, 107, 149, var(--tw-text-opacity));
}
```

but it will never be :

```css
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87, 107, 149, 1);
}
```

-------------

> Wouldn't that end up with a wrong render result if --opacity is overridden by another selector? Since we have color: rgba(87, 107, 149, 1) instead of color: rgba(87, 107, 149, var(--opacity)). 

This is a feature not a bug :)

Your source has a `var()`, removing it would alter your source code.

```css
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87, 107, 149, var(--tw-text-opacity));
}
```

This snippet in modern browsers is dynamic and setting a different value for `var(--tw-text-opacity)` will have an effect. If one of our plugins removed this behaviour it would be a bug, not a feature.

----------------

There is no practical difference (as far as I know) between these two :

```css
/* A */
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87, 107, 149, 1);
  color: rgba(87, 107, 149, var(--tw-text-opacity));
}
```

```css
/* B */
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87, 107, 149, 1);
}

@supports (top: var(--f)) {
  .text-link {
    color: rgba(87, 107, 149, var(--tw-text-opacity));
  }
}
```

Is there an example where the last two snippets do not produce the same style?
Maybe I am overlooking something :)

----------

It is also important to keep in mind that these are fallback values, to be used as graceful degradation.

Anything with custom properties is very difficult to handle because only in a browser do you know the real values.

Our goal is to set fallback values that are known to be correct.
It is not our goal to try and bring the client side behavior of dynamic custom properties to a build tool.

**Comment by satouriko** [comment]
There's sure a difference.

I created an example website: https://valuable-powder.surge.sh

On modern browsers (expected): 
<img width="460" alt="image" src="https://github.com/csstools/postcss-plugins/assets/18373361/bdd1c256-58ff-42f6-883c-feed2ba74463">


Tested on Chrome 64 (unexpected downgrade)
<img width="430" alt="image" src="https://github.com/csstools/postcss-plugins/assets/18373361/c112d14c-babd-4c65-b789-2f668910f602">
https://live.browserstack.com/dashboard#os=OS+X&os_version=Monterey&browser=Chrome&browser_version=64.0&zoom_to_fit=true&full_screen=true&url=www.google.com&speed=1

**Comment by satouriko** [comment]
What you said a lot about custom properties is irrelevant to the issue. The issue is 
postcss-color-functional-notation, i.e `rgba(87 107 149 / var(--tw-text-opacity))` -> `rgba(87, 107, 149, var(--tw-text-opacity))` is expected to function regardless of the custom properties polyfill

**Comment by romainmenke** [comment]
Please re-read my last comment :)

I did not say there is difference at all.
I also did not say that it was impossible to construct an example that shows a difference.

----

You opened an issue and stated that you wanted this source to give a color of `rgba(87, 107, 149, 1)` in as many browsers as possible :

```css
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87 107 149 / var(--tw-text-opacity));
}
```

You explicitly stated that you wanted the opacity of `1`.

The solution I proposed gives you exactly that.

I meant that there is no practical difference between your proposed solution and ours for your specific use case.

Anything else is not an option for us :/
It is either taking the plugins in a direction we do not want or it will cause problems for many other people.

-----

If you want the dynamic nature of custom properties, as you showcased in your last example, you should not be targeting browsers without native support.

> What you said a lot about custom properties is irrelevant to the issue. 

It is extremely relevant, in fact, it is the core of the issue :)
If the `postcss-custom-properties` plugin didn't exist or wasn't included in `postcss-preset-env` you wouldn't have this issue.

We have to take the existence of this plugin into account.
It is this plugin that generates this fallback : `color: rgba(87 107 149 / 1);`.

This plugin needs to do this and it needs to be done first, before any other plugin.
Changing this would be a breaking change for millions of users.

Once `color: rgba(87 107 149 / 1);` exists as a fallback, it no longer makes much sense to generate `.color: rgba(87, 107, 149, var(--tw-text-opacity));`.

**Comment by romainmenke** [comment]
Without `postcss-custom-properties` :

```css
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87 107 149 / var(--tw-text-opacity));
}
```

becomes :

```
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87, 107, 149, var(--tw-text-opacity));
}
```

**Comment by romainmenke** [comment]
Please make sure to also try this CSS in Chrome 63 :

```css
.text-link {
  --tw-text-opacity: 1;
  color: #576b95;
  color: rgba(87 107 149 / var(--tw-text-opacity));
}
```

The text then suddenly becomes black :

<img width="866" alt="Screenshot 2023-05-27 at 16 47 25" src="https://github.com/csstools/postcss-plugins/assets/11521496/872e9666-f614-44b3-899d-72f8a4dc0115">

**Comment by satouriko** [comment]
Let me try it again, I don't think you understand the issue I'm saying.

```css
/* source */
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87 107 149 / var(--tw-text-opacity));
}

/* desired */
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87 107 149, 1); /* <- this 1 is fallback and might not be correct */
  color: rgba(87, 107, 149, var(--tw-text-opacity)); /* but we have comma separated, Chorme 49~64 happy */
}

/* actually get */
.text-link {
  --tw-text-opacity: 1;
  color: rgba(87 107 149, 1); /* <- this 1 is fallback and might not be correct */
  color: rgba(87 107 149 / var(--tw-text-opacity)); /* however we have space separated, Chorme 49~64 unhappy */
}
```

Why this makes difference? Because suppose there's another rule
```css
.text-link.opacity-10 {
  --tw-text-opacity: 0.1;
}
```
On browsers between the gap (e.g Chorme 49~64), the "desired" output can render opacity (0.1) correctly, while the "actual" output cannot.

**Comment by satouriko** [comment]
Chorme 49~64, they support CSS variables, they just don't support space separated syntax. Prior to postcss-color-functional-notation@0.5.2 there's no problem for those browsers at all because postcss-color-functional-notation translates space separated to comma separated, but now it doesn't

**Comment by romainmenke** [comment]
But if the fallback of variables is not desired, then why use the plugin for it?
This is where I started to misunderstand the issue :)

Either the exact value of custom properties is important and you do not want to use the plugin. Or the exact value is not that important and having a fallback on older browsers is fine.

The this fallback is applied to Chrome < 49 or Chrome < 64 shouldn't make that big a difference today in the real world.

-------

`postcss-color-functional-notation@0.5.2` fixed real issues that we do not want to bring back.

I am looking for a way forward.  
Something that fixes your issue and doesn't bring back the old issues.

------

> This plugin needs to do this and it needs to be done first, before any other plugin.
Changing this would be a breaking change for millions of users.

I said this, but I didn't test this on the branch with my proposed changes.

It seems that with my changes it becomes safe to change the order.

This changes the output to :

```css
.test-property-with-var {
	--opacity: 1;
	color: rgba(87, 107, 149, 1);
	color: rgba(87, 107, 149, var(--opacity));
}
```

When setting `{preserve: true}` for all plugins you get :

```css
.test-property-with-var {
	--opacity: 1;
	color: rgba(87, 107, 149, 1);
	color: rgba(87, 107, 149, var(--opacity));
	color: rgba(87 107 149 / 1);
}

@supports (color: rgba(0 0 0 / 0)) and (top: var(--f)) {
	.test-property-with-var {
		color: rgba(87 107 149 / var(--opacity));
	}
}
```

**Comment by satouriko** [comment]
> But if the fallback of variables is not desired, then why use the plugin for it?

To make as many as possible users have the best desired experience. I don't think it's reasonable to affect Chrome 49~64 users because of a polyfill for Chrome < 49 users. It's sometimes acceptable to have Chrome < 49 a not that good experience, but not acceptable for Chrome < 64. Chrome is just an example, what about Safari 10 & 11, they are distributed with macOS and we do have users not able to upgrade.

Glad to see if there's something that fixes issue.

**Comment by romainmenke** [comment]
Hi @satouriko,

We've just released `postcss-preset-env` version `8.4.2` with the fix for this issue.

You can try this out in the playground :

https://preset-env.cssdb.org/playground/#eyJzb3VyY2UiOiIudGV4dC1saW5rIHtcbiAgLS10dy10ZXh0LW9wYWNpdHk6IDE7XG4gIGNvbG9yOiByZ2JhKDg3IDEwNyAxNDkgLyB2YXIoLS10dy10ZXh0LW9wYWNpdHkpKTtcbn0iLCJjb25maWciOnsiYnJvd3NlcnMiOlsiPiAwLjIlIGFuZCBub3QgZGVhZCJdLCJtaW5pbXVtVmVuZG9ySW1wbGVtZW50YXRpb25zIjowLCJzdGFnZSI6MiwibG9naWNhbCI6eyJpbmxpbmVEaXJlY3Rpb24iOiJsZWZ0LXRvLXJpZ2h0IiwiYmxvY2tEaXJlY3Rpb24iOiJ0b3AtdG8tYm90dG9tIn19fQ%3D%3D
